### ISSUE TITLE
Add proxying support to the Blazor dev server

### ISSUE THREAD

**Issue by danroth27** [dotnet/aspnetcore#24606]
Add proxying support to the Blazor dev server similar to the functionality available from other frontend frameworks, particularly those based on webpack:

- Webpack: https://webpack.js.org/configuration/dev-server/#devserverproxy
- Angular: https://angular.io/guide/build#proxying-to-a-backend-server
- React: https://create-react-app.dev/docs/proxying-api-requests-in-development/
- Vue: https://cli.vuejs.org/config/#devserver-proxy

This functionality is useful when the Blazor WebAssembly app is being developed as a separate project from its backend API during development, but will be deployed to the same domain in production. A common example is a Blazor WebAssembly frontend with an Azure Functions backend.

**Comment by javiercn** [comment]
@danroth27 can we leverage YARP for this?

**Comment by danroth27** [comment]
We have proxy middleware in the aspnet/asplabs repo: https://github.com/aspnet/AspLabs/tree/master/src/Proxy. I'm not sure if YARP is using that or not. @Tratcher?

We also have some proxying logic in SpaServices.Extensions that I believe @SteveSandersonMS wrote.

**Comment by javiercn** [comment]
I believe the main things here are to rewrite/forward requests and rewrite things like cookies and things like that?

**Comment by Tratcher** [comment]
@danroth27 YARP was "inspired-by" that lab, the code isn't very similar anymore. The relevant part is here:
https://github.com/microsoft/reverse-proxy/blob/master/src/ReverseProxy/Service/Proxy/HttpProxy.cs

**Comment by danroth27** [comment]
> I believe the main things here are to rewrite/forward requests and rewrite things like cookies and things like that?

That sounds right. The goal is to emulate single origin as much as is possible/reasonable.

**Comment by Tratcher** [comment]
Just be aware of all the things proxies break.
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-3.1

Things used for url generation, redirects, cookies, etc.:
- Scheme, TLS termination
- Host name
- Path Base

Rewriting response headers is one thing, trying to rewrite anything in the content is a very different problem.

**Comment by aaronpowell** [comment]
A quick win could be to extend the `DevServer` to use the `SpaServices.Extensions` middleware as it already contains a proxy feature: https://github.com/dotnet/aspnetcore/tree/master/src/Middleware/SpaServices.Extensions/src/Proxying

To make it interoperable with the way many JavaScript projects work, it could leverage the BrowserSync config proxy structure - https://www.browsersync.io/docs/options#option-proxy

So have it that when the devserver starts up, if it finds a `bs-config.json` file, then add the middleware and parse the proxy information out of it to setup the middleware.

**Comment by javiercn** [comment]
I think that the spa services proxy works in combination with the different JS proxies. I think we would no take a dependency on node JS for something like that.

**Comment by aaronpowell** [comment]
It doesn't appear that it uses anything from Node, the code for the proxy implementation is in https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/SpaServices.Extensions/src/Proxying/SpaProxy.cs and it added middleware to each request (https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/SpaServices.Extensions/src/Proxying/SpaProxyingExtensions.cs#L77-L82) and uses `HttpClient` to forward the request (https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/SpaServices.Extensions/src/Proxying/SpaProxy.cs#L76-L94).

It looks like prior to .NET 5 the extensions depends on SpaServices, which in turn depends on NodeServices, but if I understand the csproj correctly it's being removed in 5 RTM: (https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/SpaServices.Extensions/src/Microsoft.AspNetCore.SpaServices.Extensions.csproj#L18-L23)

**Comment by javiercn** [comment]
@aaronpowell yes, but it is proxying that request to the angular/react/webpack proxy that is started by ng run and so on, which in turn are the ones that take care of proxying API requests and things like that.

**Comment by aaronpowell** [comment]
Isn't it proxying any request that can't be handled by the ASP.Net server to an endpoint that's running whatever the endpoint runs?

**Comment by javiercn** [comment]
Now I'm not sure we are talking about the same things. Proxying to a backend server in angular/react and other things like that are handled by nodejs middlware as Dan points out.

From what I've played with it, that middleware does more than what our current proxy middleware does (rewrites cookies, etc.) that's why I'm saying the SPA proxy is not enough, but I might be wrong.

@SteveSandersonMS will know better since he wrote the thing :).

**Comment by aaronpowell** [comment]
My understanding is that the goal is to have it so that if you hit `http://localhost:5000/api/product` and there isn't a Blazor route that matches that it should then pass through to whatever you've designated `/api/product` to map to in the proxy config (it could be a local Azure Functions, it could be an asp.net core server, it could be something running on another machine), emulating single origin as Dan mentions.

**Comment by SteveSandersonMS** [comment]
@aaronpowell Makes sense!

I guess we would need to extend the framework's `MapFallback` logic so that, if an incoming request does *not* have `Accept: text/html`, then the "fallback" behavior switches over to doing a proxied request instead of returning the usual fallback response (e.g., the `index.html` file).

Alternatively we'd create some other middleware the dev server plugs in before `MapFallback`, which does this proxying for all requests that don't have `Accept: text/html`, assuming that the static-files middleware would already have handled the request if it matched a physical file or static web asset.

As for the way the proxying is achieved, it's probably ideal to use YARP if we can, because that's the more modern and better maintained implementation. As per @Tratcher's comment, there are all kinds of caveats about what can actually be proxied and how it impacts behavior, but that's the same for all the other SPA tooling proxies too.

I'm not certain what aspects of request/response rewriting are involved (besides changing the `Origin` and rewriting CORS headers). Most likely there are lots of details that aren't obvious up front. We would want to research exactly what the other SPA frameworks do to get an initial estimate of the required feature set. It might end up being a quick and cheap feature, or a hugely expensive long tail of issues.

**Comment by javiercn** [comment]
Most of the FE frameworks use something along the lines of https://webpack.js.org/configuration/dev-server/#devserverproxy, which is why I suggested our SPA proxy wasn't enough. This is also based on my experimentation of switching the SPA proxy to use the native version.

**Comment by jbasinger** [comment]
What is you allowed a string of URLs as a configuration option to the DevServer and a flag to use CORs with that list of strings?

**Comment by DontEatRice** [comment]
For inspiration, I would also add Vite's server.proxy config setting https://vite.dev/config/server-options.html#server-proxy

**Comment by robrich** [comment]
Perhaps a section in `blazor-devserversettings.json` could specify proxy info?  Maybe YARP is pulled into https://github.com/dotnet/dotnet/blob/main/src/aspnetcore/src/Components/WebAssembly/DevServer/src/Server/Program.cs if such configuration exists?

**Comment by danroth27** [comment]
Resolved in .NET 11 Preview 6. The dev-time Blazor Gateway — which replaced the WebAssembly DevServer for standalone Blazor WebAssembly apps in Preview 5 — can now proxy the app's HTTP calls to backend services through a built-in YARP reverse proxy, so the client calls the backend same-origin with no CORS during development.

- Release notes: https://github.com/dotnet/core/blob/main/release-notes/11.0/preview/preview6/aspnetcore.md#blazor-gateway-can-proxy-backend-api-calls
- Sample: https://github.com/danroth27/AspNetCore11Samples

Feedback on making the proxy configuration read from the app project's own config files is tracked in #67887.
